File: c_sharp/src/org/ldk/structs/OutputSpender.cs

using org.ldk.impl;
using org.ldk.enums;
using org.ldk.util;
using System;

namespace org { namespace ldk { namespace structs {



/** An implementation of OutputSpender */
public interface OutputSpenderInterface {
	/**Creates a [`Transaction`] which spends the given descriptors to the given outputs, plus an
	 * output to the given change destination (if sufficient change value remains). The
	 * transaction will have a feerate, at least, of the given value.
	 * 
	 * The `locktime` argument is used to set the transaction's locktime. If `None`, the
	 * transaction will have a locktime of 0. It it recommended to set this to the current block
	 * height to avoid fee sniping, unless you have some specific reason to use a different
	 * locktime.
	 * 
	 * Returns `Err(())` if the output value is greater than the input value minus required fee,
	 * if a descriptor was duplicated, or if an output descriptor `script_pubkey`
	 * does not match the one we can spend.
	 */
	Result_TransactionNoneZ spend_spendable_outputs(SpendableOutputDescriptor[] descriptors, TxOut[] outputs, byte[] change_destination_script, int feerate_sat_per_1000_weight, Option_u32Z locktime);
}

/**
 * A trait that describes a wallet capable of creating a spending [`Transaction`] from a set of
 * [`SpendableOutputDescriptor`]s.
 */
public class OutputSpender : CommonBase {
	internal bindings.LDKOutputSpender bindings_instance;
	internal long instance_idx;

	internal OutputSpender(object _dummy, long ptr) : base(ptr) { bindings_instance = null; }
	~OutputSpender() {
		if (ptr != 0) { bindings.OutputSpender_free(ptr); }
	}

	private class LDKOutputSpenderHolder { internal OutputSpender held; }
	private class LDKOutputSpenderImpl : bindings.LDKOutputSpender {
		internal LDKOutputSpenderImpl(OutputSpenderInterface arg, LDKOutputSpenderHolder impl_holder) { this.arg = arg; this.impl_holder = impl_holder; }
		private OutputSpenderInterface arg;
		private LDKOutputSpenderHolder impl_holder;
		public long spend_spendable_outputs(long _descriptors, long _outputs, long _change_destination_script, int _feerate_sat_per_1000_weight, long _locktime) {
			int _descriptors_conv_27_len = InternalUtils.getArrayLength(_descriptors);
			SpendableOutputDescriptor[] _descriptors_conv_27_arr = new SpendableOutputDescriptor[_descriptors_conv_27_len];
			for (int b = 0; b < _descriptors_conv_27_len; b++) {
				long _descriptors_conv_27 = InternalUtils.getU64ArrayElem(_descriptors, b);
				org.ldk.structs.SpendableOutputDescriptor _descriptors_conv_27_hu_conv = org.ldk.structs.SpendableOutputDescriptor.constr_from_ptr(_descriptors_conv_27);
				if (_descriptors_conv_27_hu_conv != null) { _descriptors_conv_27_hu_conv.ptrs_to.AddLast(this); };
				_descriptors_conv_27_arr[b] = _descriptors_conv_27_hu_conv;
			}
			bindings.free_buffer(_descriptors);
			int _outputs_conv_7_len = InternalUtils.getArrayLength(_outputs);
			TxOut[] _outputs_conv_7_arr = new TxOut[_outputs_conv_7_len];
			for (int h = 0; h < _outputs_conv_7_len; h++) {
				long _outputs_conv_7 = InternalUtils.getU64ArrayElem(_outputs, h);
				TxOut _outputs_conv_7_conv = new TxOut(null, _outputs_conv_7);
				_outputs_conv_7_arr[h] = _outputs_conv_7_conv;
			}
			bindings.free_buffer(_outputs);
			byte[] _change_destination_script_conv = InternalUtils.decodeUint8Array(_change_destination_script);
			org.ldk.structs.Option_u32Z _locktime_hu_conv = org.ldk.structs.Option_u32Z.constr_from_ptr(_locktime);
			if (_locktime_hu_conv != null) { _locktime_hu_conv.ptrs_to.AddLast(this); };
			Result_TransactionNoneZ ret = arg.spend_spendable_outputs(_descriptors_conv_27_arr, _outputs_conv_7_arr, _change_destination_script_conv, _feerate_sat_per_1000_weight, _locktime_hu_conv);
				GC.KeepAlive(arg);
			long result = ret.clone_ptr();
			return result;
		}
	}

	/** Creates a new instance of OutputSpender from a given implementation */
	public static OutputSpender new_impl(OutputSpenderInterface arg) {
		LDKOutputSpenderHolder impl_holder = new LDKOutputSpenderHolder();
		LDKOutputSpenderImpl impl = new LDKOutputSpenderImpl(arg, impl_holder);
		long[] ptr_idx = bindings.LDKOutputSpender_new(impl);

		impl_holder.held = new OutputSpender(null, ptr_idx[0]);
		impl_holder.held.instance_idx = ptr_idx[1];
		impl_holder.held.bindings_instance = impl;
		return impl_holder.held;
	}

	/**
	 * Creates a [`Transaction`] which spends the given descriptors to the given outputs, plus an
	 * output to the given change destination (if sufficient change value remains). The
	 * transaction will have a feerate, at least, of the given value.
	 * 
	 * The `locktime` argument is used to set the transaction's locktime. If `None`, the
	 * transaction will have a locktime of 0. It it recommended to set this to the current block
	 * height to avoid fee sniping, unless you have some specific reason to use a different
	 * locktime.
	 * 
	 * Returns `Err(())` if the output value is greater than the input value minus required fee,
	 * if a descriptor was duplicated, or if an output descriptor `script_pubkey`
	 * does not match the one we can spend.
	 */
	public Result_TransactionNoneZ spend_spendable_outputs(SpendableOutputDescriptor[] descriptors, TxOut[] outputs, byte[] change_destination_script, int feerate_sat_per_1000_weight, org.ldk.structs.Option_u32Z locktime) {
		long ret = bindings.OutputSpender_spend_spendable_outputs(this.ptr, InternalUtils.encodeUint64Array(InternalUtils.mapArray(descriptors, descriptors_conv_27 => descriptors_conv_27.ptr)), InternalUtils.encodeUint64Array(InternalUtils.mapArray(outputs, outputs_conv_7 => outputs_conv_7.ptr)), InternalUtils.encodeUint8Array(change_destination_script), feerate_sat_per_1000_weight, locktime.ptr);
		GC.KeepAlive(this);
		GC.KeepAlive(descriptors);
		GC.KeepAlive(outputs);
		GC.KeepAlive(change_destination_script);
		GC.KeepAlive(feerate_sat_per_1000_weight);
		GC.KeepAlive(locktime);
		if (ret >= 0 && ret <= 4096) { return null; }
		Result_TransactionNoneZ ret_hu_conv = Result_TransactionNoneZ.constr_from_ptr(ret);
		foreach (SpendableOutputDescriptor descriptors_conv_27 in descriptors) { if (this != null) { this.ptrs_to.AddLast(descriptors_conv_27); }; };
		if (this != null) { this.ptrs_to.AddLast(locktime); };
		return ret_hu_conv;
	}

}
} } }

File: src/main/java/org/ldk/structs/OutputSpender.java
package org.ldk.structs;

import org.ldk.impl.bindings;
import org.ldk.enums.*;
import org.ldk.util.*;
import java.util.Arrays;
import java.lang.ref.Reference;
import javax.annotation.Nullable;

/**
 * A trait that describes a wallet capable of creating a spending [`Transaction`] from a set of
 * [`SpendableOutputDescriptor`]s.
 */
@SuppressWarnings("unchecked") // We correctly assign various generic arrays
public class OutputSpender extends CommonBase {
	final bindings.LDKOutputSpender bindings_instance;
	OutputSpender(Object _dummy, long ptr) { super(ptr); bindings_instance = null; }
	private OutputSpender(bindings.LDKOutputSpender arg) {
		super(bindings.LDKOutputSpender_new(arg));
		this.ptrs_to.add(arg);
		this.bindings_instance = arg;
	}
	@Override @SuppressWarnings("deprecation")
	protected void finalize() throws Throwable {
		if (ptr != 0) { bindings.OutputSpender_free(ptr); } super.finalize();
	}
	/**
	 * Destroys the object, freeing associated resources. After this call, any access
	 * to this object may result in a SEGFAULT or worse.
	 *
	 * You should generally NEVER call this method. You should let the garbage collector
	 * do this for you when it finalizes objects. However, it may be useful for types
	 * which represent locks and should be closed immediately to avoid holding locks
	 * until the GC runs.
	 */
	public void destroy() {
		if (ptr != 0) { bindings.OutputSpender_free(ptr); }
		ptr = 0;
	}
	public static interface OutputSpenderInterface {
		/**
		 * Creates a [`Transaction`] which spends the given descriptors to the given outputs, plus an
		 * output to the given change destination (if sufficient change value remains). The
		 * transaction will have a feerate, at least, of the given value.
		 * 
		 * The `locktime` argument is used to set the transaction's locktime. If `None`, the
		 * transaction will have a locktime of 0. It it recommended to set this to the current block
		 * height to avoid fee sniping, unless you have some specific reason to use a different
		 * locktime.
		 * 
		 * Returns `Err(())` if the output value is greater than the input value minus required fee,
		 * if a descriptor was duplicated, or if an output descriptor `script_pubkey`
		 * does not match the one we can spend.
		 */
		Result_TransactionNoneZ spend_spendable_outputs(SpendableOutputDescriptor[] descriptors, TxOut[] outputs, byte[] change_destination_script, int feerate_sat_per_1000_weight, Option_u32Z locktime);
	}
	private static class LDKOutputSpenderHolder { OutputSpender held; }
	public static OutputSpender new_impl(OutputSpenderInterface arg) {
		final LDKOutputSpenderHolder impl_holder = new LDKOutputSpenderHolder();
		impl_holder.held = new OutputSpender(new bindings.LDKOutputSpender() {
			@Override public long spend_spendable_outputs(long[] descriptors, long[] outputs, byte[] change_destination_script, int feerate_sat_per_1000_weight, long locktime) {
				int descriptors_conv_27_len = descriptors.length;
				SpendableOutputDescriptor[] descriptors_conv_27_arr = new SpendableOutputDescriptor[descriptors_conv_27_len];
				for (int b = 0; b < descriptors_conv_27_len; b++) {
					long descriptors_conv_27 = descriptors[b];
					org.ldk.structs.SpendableOutputDescriptor descriptors_conv_27_hu_conv = org.ldk.structs.SpendableOutputDescriptor.constr_from_ptr(descriptors_conv_27);
					if (descriptors_conv_27_hu_conv != null) { descriptors_conv_27_hu_conv.ptrs_to.add(this); };
					descriptors_conv_27_arr[b] = descriptors_conv_27_hu_conv;
				}
				int outputs_conv_7_len = outputs.length;
				TxOut[] outputs_conv_7_arr = new TxOut[outputs_conv_7_len];
				for (int h = 0; h < outputs_conv_7_len; h++) {
					long outputs_conv_7 = outputs[h];
					TxOut outputs_conv_7_conv = new TxOut(null, outputs_conv_7);
					outputs_conv_7_arr[h] = outputs_conv_7_conv;
				}
				org.ldk.structs.Option_u32Z locktime_hu_conv = org.ldk.structs.Option_u32Z.constr_from_ptr(locktime);
				if (locktime_hu_conv != null) { locktime_hu_conv.ptrs_to.add(this); };
				Result_TransactionNoneZ ret = arg.spend_spendable_outputs(descriptors_conv_27_arr, outputs_conv_7_arr, change_destination_script, feerate_sat_per_1000_weight, locktime_hu_conv);
				Reference.reachabilityFence(arg);
				long result = ret.clone_ptr();
				return result;
			}
		});
		return impl_holder.held;
	}
	/**
	 * Creates a [`Transaction`] which spends the given descriptors to the given outputs, plus an
	 * output to the given change destination (if sufficient change value remains). The
	 * transaction will have a feerate, at least, of the given value.
	 * 
	 * The `locktime` argument is used to set the transaction's locktime. If `None`, the
	 * transaction will have a locktime of 0. It it recommended to set this to the current block
	 * height to avoid fee sniping, unless you have some specific reason to use a different
	 * locktime.
	 * 
	 * Returns `Err(())` if the output value is greater than the input value minus required fee,
	 * if a descriptor was duplicated, or if an output descriptor `script_pubkey`
	 * does not match the one we can spend.
	 */
	public Result_TransactionNoneZ spend_spendable_outputs(SpendableOutputDescriptor[] descriptors, TxOut[] outputs, byte[] change_destination_script, int feerate_sat_per_1000_weight, org.ldk.structs.Option_u32Z locktime) {
		long ret = bindings.OutputSpender_spend_spendable_outputs(this.ptr, descriptors != null ? Arrays.stream(descriptors).mapToLong(descriptors_conv_27 -> descriptors_conv_27.ptr).toArray() : null, outputs != null ? Arrays.stream(outputs).mapToLong(outputs_conv_7 -> outputs_conv_7.ptr).toArray() : null, change_destination_script, feerate_sat_per_1000_weight, locktime.ptr);
		Reference.reachabilityFence(this);
		Reference.reachabilityFence(descriptors);
		Reference.reachabilityFence(outputs);
		Reference.reachabilityFence(change_destination_script);
		Reference.reachabilityFence(feerate_sat_per_1000_weight);
		Reference.reachabilityFence(locktime);
		if (ret >= 0 && ret <= 4096) { return null; }
		Result_TransactionNoneZ ret_hu_conv = Result_TransactionNoneZ.constr_from_ptr(ret);
		return ret_hu_conv;
	}

}
